### PR TITLE
Yoda: `MatrixEntry` type + user function call to access underlying `Container`

### PR DESCRIPTION
**Pull request by mzhurovich**
@dkorolev PTAL

### CODE REVIEW

File: yoda/exceptions.h [hunk 1]
```diff
@@ -55,6 +57,35 @@ struct KeyAlreadyExistsException : KeyAlreadyExistsCoverException {
...
+struct CellAlreadyExistsCoverException : bricks::Exception {};
+
+template <typename ENTRY>
+struct CellAlreadyExistsException : CellAlreadyExistsCoverException {
+  typedef ENTRY T_ENTRY;
+  const ENTRY entry;
+  explicit CellAlreadyExistsException(const ENTRY& entry) : entry(entry) {}
+};
+
+// Exceptions used for both, `KeyEntry` and `MatrixEntry` types.
```
Comment (dkorolev) [inline yoda/exceptions.h]: Redundant comma after `both`.

File: yoda/keyvalue.h [hunk 2]
```diff
@@ -0,0 +1,378 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+          (c) 2015 Maxim Zhurovich <zhurovich@gmail.com>
```
Comment (dkorolev) [inline yoda/keyvalue.h]: Flip the order?

File: yoda/matrix.h [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+namespace yoda {
+
+// Here the data is stored. 
+template <typename ENTRY>
+struct Container<MatrixEntry<ENTRY>> {
+  typedef ENTRY T_ENTRY;
+  typedef ENTRY_ROW_TYPE<T_ENTRY> T_ROW;
+  typedef ENTRY_COL_TYPE<T_ENTRY> T_COL;
+
+  T_MAP_TYPE<T_ROW, T_MAP_TYPE<T_COL, T_ENTRY>> forward;
```
Comment (dkorolev) [inline yoda/matrix.h:50]: :+1:

File: yoda/matrix.h [hunk 4]
```diff
@@ -0,0 +1,344 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
```
Comment (dkorolev) [inline yoda/matrix.h]: Definitely flip the names :-)

File: yoda/keyvalue.h [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+  void AsyncAdd(const T_ENTRY& entry,
+                T_VOID_CALLBACK on_success,
+                T_VOID_CALLBACK on_failure = [](const T_KEY&) {}) {
+    mq_.EmplaceMessage(new MQMessageAdd(entry, on_success, on_failure));
+  }
+
+  void Add(const T_ENTRY& entry) { AsyncAdd(entry).get(); }
+
+  void ApplyFunction(T_USER_FUNCTION function) {
```
Comment (dkorolev) [inline yoda/keyvalue.h]: `Async*` prefix here and below?

File: yoda/test.cc [hunk 6]
```diff
@@ -226,4 +229,163 @@ TEST(Sherlock, NonPolymorphicKeyValueStorage) {
   api.Subscribe(listener).Join();
   EXPECT_EQ(data.seen_, 6u);
   EXPECT_EQ("2=0.50,3=0.33,4=0.25,5=0.20,6=0.17,7=0.76", data.results_);
+
+  // Test user function accessing the underlying container.
+  typedef yoda::Container<yoda::KeyEntry<KeyValueEntry>> T_CONTAINER;
+  double sum = 0.0;
+  bool done = false;
+  api.ApplyFunction([&](const T_CONTAINER& container) {
+    for (auto i : container.data) {
```
Comment (dkorolev) [inline yoda/test.cc]: `const` `auto`

File: yoda/test.cc [hunk 7]
```diff
@@ -226,4 +229,163 @@ TEST(Sherlock, NonPolymorphicKeyValueStorage) {
...
+    done = true;
+  });
+  while (!done) {
+    ;  // Spin lock.
+  }
+  EXPECT_DOUBLE_EQ(2.21, sum);
+}
+
+struct MatrixCell {
+  size_t row;
```
Comment (dkorolev) [inline yoda/test.cc:250]: Yay! :+1:

File: yoda/test.cc [hunk 8]
```diff
@@ -226,4 +229,163 @@ TEST(Sherlock, NonPolymorphicKeyValueStorage) {
...
+  while (!cbt4.called) {
+    ;  // Spin lock.
+  }
+
+   // Test user function accessing the underlying container.
+  typedef yoda::Container<yoda::MatrixEntry<MatrixCell>> T_CONTAINER;
+  size_t row_index_sum = 0;
+  int value_sum = 0;
+  bool done = false;
+  api.ApplyFunction([&](const T_CONTAINER& container) {
```
Comment (dkorolev) [inline yoda/test.cc]: :+1:

File: yoda/test.cc [hunk 9]
```diff
@@ -226,4 +229,163 @@ TEST(Sherlock, NonPolymorphicKeyValueStorage) {
...
+    // Testing forward and transposed matrices.
+    for (auto rit : container.forward) {
```
Comment (dkorolev) [inline yoda/test.cc]: `const` auto here and below.

File: yoda/types.h [hunk 10]
```diff
@@ -104,6 +123,112 @@ void SetKey(T_ENTRY& entry, ENTRY_KEY_TYPE<T_ENTRY> key) {
...
+  return true;
+}
+
+template <typename T_ENTRY, bool HAS_ROW_FUNCTION>
+struct ROW_ACCESSOR_IMPL {};
+
+template <typename T_ENTRY>
+struct ROW_ACCESSOR_IMPL<T_ENTRY, false> {
+  typedef decltype(std::declval<T_ENTRY>().row) T_ROW;
+  static typename std::conditional<std::is_pod<T_ROW>::value, T_ROW, const T_ROW&>::type GetRow(
```
Comment (dkorolev) [inline yoda/types.h:146]: Note to self: We should probably move this `is_pod` conditional into `Bricks`.

File: yoda/types.h [hunk 11]
```diff
@@ -104,6 +123,112 @@ void SetKey(T_ENTRY& entry, ENTRY_KEY_TYPE<T_ENTRY> key) {
...
+using ROW_ACCESSOR = ROW_ACCESSOR_IMPL<T_ENTRY, HasRowFunction<T_ENTRY>(0)>;
+
+template <typename T_ENTRY>
+typename ROW_ACCESSOR<T_ENTRY>::T_ROW GetRow(const T_ENTRY& entry) {
+  return ROW_ACCESSOR<T_ENTRY>::GetRow(entry);
+}
+
+template <typename T_ENTRY>
+using ENTRY_ROW_TYPE =
+    typename std::remove_cv<typename std::remove_reference<typename ROW_ACCESSOR<T_ENTRY>::T_ROW>::type>::type;
```
Comment (dkorolev) [inline yoda/types.h:173]: Note to self: Make use of the shortened versions of this, which I am just about to push to `Bricks`!

**Comment by dkorolev** [comment]
Yay! :+1: @mzhurovich, your code is, as always, clean and efficient!

Extra bonus points for `Apply` functions and extensive test, and extra extra kudos for following file structure that is not pushed yet, but will be becoming live soon.

The merge order will be mine then yours, and Dima is the merge tool. Looking forward to syncing up in ~11 hours. I'll do my best to make my change polished and sent your way in the next several hours, for you to have enough time to look at it by the time we sync up.

Thanks!
Dima

**Comment by mzhurovich** [comment]
@dkorolev renamed `ApplyFunction` to `AsyncCallFunction`

**Comment by dkorolev** [comment]
@mzhurovich :+1:, the rename and all the minor changes commit. Talk soon!

**Comment by dkorolev** [comment]
@mzhurovich : This one is safe to close as https://github.com/KnowSheet/Sherlock/pull/29 contains it, right?

**Comment by mzhurovich** [comment]
@dkorolev Yep!

**Pull request closed without merging**